File: 113.c
#include <stdio.h>
#include <stdlib.h>

#define NUM_DIGITS 100

long num_decreasing( long num_digits, 
		     long first_digit,
		     long **buffer ) {
  long (*answers)[10];
  long num = 0;
  int i;
  answers = (long (*)[10])buffer;
  if( answers == NULL ) {
    /*printf( "calloc\n" ); */
    answers = (long (*)[10])calloc(sizeof(long), num_digits*20);
  }
  else if( answers[num_digits][first_digit] > 0 ) {
    //printf( "%ld\n", answers[num_digits][first_digit] );
    return answers[num_digits][first_digit];
  }  
  printf( "num_dec(%ld, %ld) = ", num_digits, first_digit );
  if( num_digits < 1 || first_digit == 0 ) {
    printf( "1\n" );
    return 1;
  }
  else if( num_digits == 1 ) {
    printf( "%ld\n", first_digit + 1 );
    return first_digit + 1;
  }
 
  printf( "....\n" );
  /* how many times current digit repeats */
  for( i=0; i<=num_digits; i++ ) {
    num += num_decreasing( num_digits - i, first_digit - 1, (long **)answers );
  }
  printf( ".......(%ld, %ld) = %ld\n", num_digits, first_digit, num );
  answers[num_digits][first_digit] = num;
  /*if( buffer == NULL )
    free( (void *)answers );*/
  return num;
}
  

void main() {
  long n=0;
  n = num_decreasing( NUM_DIGITS, 10, NULL ) +
    num_decreasing( NUM_DIGITS, 9, NULL ) - 2 - 10 * NUM_DIGITS;
  printf( "n=%ld\n", n );
}
  
/*
non_bouncy = 99
increasing = 54
decreasing = 63
both = 18
*/

/*
non_bouncy = 474
increasing = 219
decreasing = 282
both = 27
*/

/*
non_bouncy = 12951
increasing = 5004
decreasing = 8001
both = 54
*/

File: 113.hs
import Debug.Trace
import Data.Char

digi = "01234567890"

numDecreasing :: String -> Int -> Int -> Integer
numDecreasing pfx 0 0 = trace (pfx ++ "  (0,0) -- B0 -----") 
                        1
numDecreasing pfx len 0 = trace (pfx ++ (replicate len '0') ++ "  " ++ (show (len,0)) ++ " -- B1 --") 
                          ((toInteger len) + 1)--run out of digits before end of number
numDecreasing pfx 0 digits = trace (pfx ++ "  " ++ (show (0,digits)) ++ " -- B2") 
                          0 -- create the whole number
numDecreasing pfx len digits = --trace (pfx ++ "    " (show (len,digits))) $
                           ( numDecreasing (pfx ++ [digi!!digits]) (len - 1) digits ) + 
                           ( numDecreasing (pfx) len (digits - 1) )
                           -- current digit repeats n times
                           
-- numDecreasing!!maxDigit!!numberOfDigits
numDecreasing' :: [[Int]]
numDecreasing' = ([0,0..]):[ 0:d:[ (numDecreasing'!!(d)!!(num-1)) + (numDecreasing'!!(d-1)!!num) + num | num <- [2..] ] |
                                  d <- ( [1..9] :: [Int] ) ]
                 
isDecreasing n = all ((>=0).(uncurry (-))) $ p $ d where
  p [x] = []
  p (x:y:xs) = (x,y):( p (y:xs) )
  d = map digitToInt $ show n
                                                     
numFlat len = 9*len
      
numNonBouncy len = let ndec = numDecreasing "" len 10 in
  (ndec - 11)*2 + 9
  
  
notDs ds n = and $ map ($(show n)) $ map notElem $ map (head.show) ds

numDec maxd numd = length $ filter isDecreasing $ filter (notDs [maxd+1..9]) [1..(10^numd)-1]
{-
22 
21
20
11
10
02
01
00

-}